components/logins/src/error.rs

```diff
@@ -97,192 +86,71 @@ pub enum InvalidLogin {
-    NoTarget,
-    /// Login has illegal field
-    IllegalFieldValue,
-}
-
-// A port of the error conversion stuff that was in ffi.rs - it turns our
-// "internal" errors into "public" ones.
-impl From<Error> for LoginsStorageError {
-    fn from(e: Error) -> LoginsStorageError {
-        use sync15::ErrorKind as Sync15ErrorKind;
 
-        let label = e.label().to_string();
-        let kind = e.kind();
-        match kind {
-            ErrorKind::SyncAdapterError(e) => {
-                log::error!("Sync error {:?}", e);
-                match e.kind() {
-                    Sync15ErrorKind::TokenserverHttpError(401)
-                    | Sync15ErrorKind::BadKeyLength(..) => {
-                        LoginsStorageError::SyncAuthInvalid(label)
-                    }
-                    Sync15ErrorKind::RequestError(_) => LoginsStorageError::RequestFailed(label),
-                    _ => LoginsStorageError::UnexpectedLoginsStorageError(label),
-                }
-            }
-            ErrorKind::NoSuchRecord(id) => {
-                log::error!("No record exists with id {}", id);
-                LoginsStorageError::NoSuchRecord(label)
-            }
-            ErrorKind::InvalidLogin(desc) => {
-                log::warn!("Invalid login: {}", desc);
-                match desc {
-                    InvalidLogin::EmptyOrigin => {
-                        LoginsStorageError::InvalidRecord(label, InvalidLoginReason::EmptyOrigin)
-                    }
-                    InvalidLogin::EmptyPassword => {
-                        LoginsStorageError::InvalidRecord(label, InvalidLoginReason::EmptyPassword)
-                    }
-                    InvalidLogin::DuplicateLogin => {
-                        LoginsStorageError::InvalidRecord(label, InvalidLoginReason::DuplicateLogin)
-                    }
-                    InvalidLogin::BothTargets => {
-                        LoginsStorageError::InvalidRecord(label, InvalidLoginReason::BothTargets)
-                    }
-                    InvalidLogin::NoTarget => {
-                        LoginsStorageError::InvalidRecord(label, InvalidLoginReason::NoTarget)
-                    }
-                    InvalidLogin::IllegalFieldValue { .. } => LoginsStorageError::InvalidRecord(
-                        label,
-                        InvalidLoginReason::IllegalFieldValue,
-                    ),
-                }
-            }
-            // We can't destructure `err` without bringing in the libsqlite3_sys crate
-            // (and I'd really rather not) so we can't put this in the match.
-            ErrorKind::SqlError(rusqlite::Error::SqliteFailure(err, _))
+    /// Report this error to our tracking system if appropriate
+    pub fn track_error(&self) {
+        // TODO: replace this with new error reporting code:
+        //
+        //   - `log::error!` should change to `error_support::report_error!` once #4979 is merged.
```

Comment:
#4979 is merged :) So I think we can probably do that now and it would be replaced with a single line calling `report_error(self.error_type(), self.to_string())` or similar?

(hrm - I can't see where this is called though, so maybe it should just be removed?)

Reply:
I didn't add the `report_error()` call because I wanted to wait and see how the new system works for the `QueryReturnedNoRows` error before committing to the switch.

I completely forgot adding the call to `track_error()` and that's one of the bigger changes needed.  Before we were relying on the `Into<LoginsStorageError>` impl, which [UniFFI implicitly calls](https://github.com/mozilla/uniffi-rs/commit/775a5bf8ef9256eac8cbd42da6c78097de816f80).  I just pushed some code to implement this, but there were a few possible options and I'm not sure if I picked the best one.  Here are the options I could think of:
  - Wrap each `LoginStore` method in a macro -- this is what I ended up going with
  - Do the same thing with a proc-macro.  This would be more ergonomic, since we could wrap the entire `impl` block.  But it would be more work to implement.
  - Keep the `Error` -> `LoginsStorageError` conversion, but only for error tracking.  `LoginsStorageError` would have the exact same shape as `Error`.  The main reason I didn't go with this one, is because I'm not sure UniFFI should be implicitly calling `into()`